
### Supervised Learning
### Activity: Building a Student Intervention System

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail or pass. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** 

Classification.
The Classification algorithm is used to identify the category of new observations on the basis of training data.Since we need to identify two categories of early intervention of "pass" and "fail" for students,this supervised learning technique is used.

### Question-2
load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Read student data
dt=pd.read_csv("student-data.csv")

In [3]:
dt

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0    ...       no       no       4         3      4    1    1      3        6   
1    ...      yes       no       5         3      3    1    1      3        4   
2    ...      yes       no       4         3      2    2    3      3       10   
3    ...      yes      yes       3         2      2    1    1      5        2   
4    ...       no       no       4         3      2    1    2      5        4   
..   ...      ...      ...     ...       ...    ...  ...  ...    ...      ...   
390  ...       no       no       5         5      4    4    5      4       11   
391  ...      yes       no       2         4      5    3    4      2        3   
392  ...       no       no       5         5      3    3    3      3        3   
393  ...      yes       no       4         4      1    3    4      5        0   
394  ...      yes       no       3         2      3    3    3      5        5   

    passed  
0       no  
1       no  
2      yes  
3      yes  
4      yes  
..     ...  
390     no  
391    yes  
392     no  
393    yes  
394     no  

[395 rows x 31 columns]

### Question-3
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [4]:
# Calculate number of students
dt.shape

(395, 31)

In [5]:
# The total no. of rows represent the total no. of students
n_students = dt.shape[0]


In [6]:
# Calculate number of features
   # The total no. of columns represents the no. of features of each student
n_features = dt.shape[1]


In [7]:
# Calculate passing students
n_passed = dt["passed"].value_counts().yes


In [8]:
# Calculate failing students
n_failed = dt["passed"].value_counts().no


In [9]:
# Calculate graduation rate
  # graduation rate is (n_passed/n_students)*100
grad_rate = ((n_passed/n_students)*100)


In [10]:
# Print the results
print("Total number of students in the dataset :",n_students)
print("Total number of features of each student :",n_features)
print("The number of students passed :",n_passed)
print("The number of students failed :",n_failed)
print("Graduation rate of the class :",grad_rate)

Total number of students in the dataset : 395
Total number of features of each student : 31
The number of students passed : 265
The number of students failed : 130
Graduation rate of the class : 67.08860759493672


## Preparing the Data
you will prepare the data for modeling, training and testing.

### Question-4 Identify feature and target columns


separate the student data into feature and target columns to see if any features are non-numeric.

In [11]:
dt.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'passed'],
      dtype='object')

In [12]:
# Extract feature columns
feat = dt[dt.columns[0:30]]
feat

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0    ...    yes       no        no       4         3     4    1    1      3   
1    ...    yes      yes        no       5         3     3    1    1      3   
2    ...    yes      yes        no       4         3     2    2    3      3   
3    ...    yes      yes       yes       3         2     2    1    1      5   
4    ...    yes       no        no       4         3     2    1    2      5   
..   ...    ...      ...       ...     ...       ...   ...  ...  ...    ...   
390  ...    yes       no        no       5         5     4    4    5      4   
391  ...    yes      yes        no       2         4     5    3    4      2   
392  ...    yes       no        no       5         5     3    3    3      3   
393  ...    yes      yes        no       4         4     1    3    4      5   
394  ...    yes      yes        no       3         2     3    3    3      5   

    absences  
0          6  
1          4  
2         10  
3          2  
4          4  
..       ...  
390       11  
391        3  
392        3  
393        0  
394        5  

[395 rows x 30 columns]

In [13]:
# Extract target column 'passed'
targ = dt[dt.columns[30:31]]
targ

passed
0       no
1       no
2      yes
3      yes
4      yes
..     ...
390     no
391    yes
392     no
393    yes
394     no

[395 rows x 1 columns]

In [14]:
# Separate the data into feature data and target data (X and y, respectively)

In [15]:
X = feat
y= targ

### Question-5 Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [16]:
# label encoding
from sklearn import preprocessing
LabelEncoder = preprocessing.LabelEncoder()

cols_le=["school","sex","address","famsize","Pstatus","schoolsup","famsup","paid","activities","nursery","higher",
         "internet","romantic"]
X[cols_le]=X[cols_le].apply(LabelEncoder.fit_transform)
    

X.head()

C:\Users\Acer\anaconda3\lib\site-packages\pandas\core\frame.py:3191: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


school  sex  age  address  famsize  Pstatus  Medu  Fedu     Mjob      Fjob  \
0       0    0   18        1        0        0     4     4  at_home   teacher   
1       0    0   17        1        0        1     1     1  at_home     other   
2       0    0   15        1        1        1     1     1  at_home     other   
3       0    0   15        1        0        1     4     2   health  services   
4       0    0   16        1        0        1     3     3    other     other   

   ... higher internet  romantic  famrel  freetime  goout  Dalc  Walc  health  \
0  ...      1        0         0       4         3      4     1     1       3   
1  ...      1        1         0       5         3      3     1     1       3   
2  ...      1        1         0       4         3      2     2     3       3   
3  ...      1        1         1       3         2      2     1     1       5   
4  ...      1        0         0       4         3      2     1     2       5   

   absences  
0         6  
1         4  
2        10  
3         2  
4         4  

[5 rows x 30 columns]

In [17]:
cols_ohe = ["guardian","reason","Fjob","Mjob"]

X= pd.get_dummies(X,columns = cols_ohe)

X.head()



school  sex  age  address  famsize  Pstatus  Medu  Fedu  traveltime  \
0       0    0   18        1        0        0     4     4           2   
1       0    0   17        1        0        1     1     1           1   
2       0    0   15        1        1        1     1     1           1   
3       0    0   15        1        0        1     4     2           1   
4       0    0   16        1        0        1     3     3           1   

   studytime  ...  Fjob_at_home  Fjob_health  Fjob_other  Fjob_services  \
0          2  ...             0            0           0              0   
1          2  ...             0            0           1              0   
2          2  ...             0            0           1              0   
3          3  ...             0            0           0              1   
4          2  ...             0            0           1              0   

   Fjob_teacher  Mjob_at_home  Mjob_health  Mjob_other  Mjob_services  \
0             1             1            0           0              0   
1             0             1            0           0              0   
2             0             1            0           0              0   
3             0             0            1           0              0   
4             0             0            0           1              0   

   Mjob_teacher  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 43 columns]

### Question - 6 Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. you will need to implement the following:
- Randomly shuffle and split the data (`X`, `y`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [18]:
# splitting the data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25,random_state=42)

In [19]:
# Show the results of the split
X_train.shape

(296, 43)

In [20]:
y_train.shape

(296, 1)

### Question - 7  Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data and measure the accuracy score.

###  Model Application
*List three supervised learning models that are appropriate for this problem. What are the general applications of each model? What are their strengths and weaknesses? Given what you know about the data, why did you choose these models to be applied?*

In [21]:
# Import the three supervised learning models from sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

## Logistic regression

** Advantages
* It is one of the simplest machine learning algorithms which does not require high computational power.
* It can be be used to find out the relationship between features.
* It is used as a benchmark model to measure performance,as it is relatively quick.Here the targert of performance of students    can be done quickly.

** Disadvantages
* On high dimensional datasets,it leads to overfitting on training sets. Here the dataset being low dimensional it is less     prone to overfitting.
* It is difficult to capture complex relationships. This dataset is devoid of such complexity it can be used.
* Only important and relevant features should be used.Otherwise the model's predictive value may degrade.So here we taken only the relevant features to obtain good accuracy.

In [22]:
# fit model-1 on training data

In [23]:
lr = LogisticRegression()  
lr.fit(X_train,  y_train)

C:\Users\Acer\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\Users\Acer\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [24]:
# predict on test data

In [25]:
y_predict = lr.predict(X_test) 

In [26]:
# calculate the accuracy score

In [27]:
print("Accuracy_score is",round(accuracy_score(y_test,y_predict),2))

Accuracy_score is 0.7


## SVM Linear

** Advantages

* SVM works relatively well when there is a clear margin of separation between classes.In this dataset the target is a pass or fail of students with reference to the features where a clear cut margin of class is defined.
* SVM is relatively memory efficient

** Disadvantages

* SVM algorithm is not suitable for large dataset. As the given dataset is not large this can be applied.
* SVM does not perform well when dataset has more noise i.e, target classes are overlapping.Since the given data set has a clear cut margin,therefore no issue of target class overlapping is observed.
  

In [28]:
# fit model-2  on traning data 

In [29]:
svm = SVC(probability=True, kernel="linear")  
svm.fit(X_train,  y_train) 

C:\Users\Acer\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


SVC(kernel='linear', probability=True)

In [30]:
# predict on the test data 

In [31]:
y_predict = svm.predict(X_test)

In [32]:
# calculate the accuracy score

In [33]:
print("Accuracy_score is:",round(accuracy_score(y_test,y_predict),2))

Accuracy_score is: 0.69


## Random Forest 

** Advantages
* It can handle missing values automatically.
* It is usually robust to outliers and can handle them automatically.
* It can be used to solve both classification as well as regression problems.

** Disadvantages
* It requires much more time to train as compared to decision trees as it generates lot of trees.
* Since it  creates a lot of trees, the algorithm requires much more computational power and resources.


In [34]:
# fit model-3 on traning data 

In [35]:
rf = RandomForestClassifier() 
rf.fit(X_train,  y_train) 

<ipython-input-35-377e9ccf0ee4>:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(X_train,  y_train) #Train


RandomForestClassifier()

In [36]:
# predict on the test data 

In [37]:
y_predict = rf.predict(X_test)  

In [38]:
# calculate the accuracy score

In [39]:
print("Accuracy_score is:",round(accuracy_score(y_test,y_predict),2))


Accuracy_score is: 0.65


### Checking the feature importance of Random Forest

In [41]:
pd.Series(rf.feature_importances_,index=X.columns).sort_values(ascending=False)

absences             0.093231
failures             0.064319
goout                0.058175
age                  0.049480
Medu                 0.043088
freetime             0.040570
famrel               0.039434
Walc                 0.037000
Fedu                 0.035863
health               0.035104
studytime            0.033995
Dalc                 0.028477
traveltime           0.024186
paid                 0.021599
sex                  0.020480
reason_course        0.020143
higher               0.020021
schoolsup            0.019445
activities           0.018164
Mjob_other           0.017429
famsup               0.016823
famsize              0.016751
Mjob_at_home         0.015585
reason_reputation    0.014609
guardian_mother      0.014514
Fjob_services        0.014048
romantic             0.014019
nursery              0.013987
guardian_father      0.013929
internet             0.013443
address              0.013215
Fjob_other           0.013157
Mjob_services        0.012921
reason_hom

### Feature reduction

In [42]:
X = X.drop(["Fjob_health","Fjob_teacher","school","Pstatus","reason_other","Fjob_at_home","Mjob_health","Mjob_teacher"],axis=1)

In [43]:
X.columns

Index(['sex', 'age', 'address', 'famsize', 'Medu', 'Fedu', 'traveltime',
       'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime',
       'goout', 'Dalc', 'Walc', 'health', 'absences', 'guardian_father',
       'guardian_mother', 'guardian_other', 'reason_course', 'reason_home',
       'reason_reputation', 'Fjob_other', 'Fjob_services', 'Mjob_at_home',
       'Mjob_other', 'Mjob_services'],
      dtype='object')

In [45]:
X.describe()

sex         age     address     famsize        Medu        Fedu  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     0.473418   16.696203    0.777215    0.288608    2.749367    2.521519   
std      0.499926    1.276043    0.416643    0.453690    1.094735    1.088201   
min      0.000000   15.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000   16.000000    1.000000    0.000000    2.000000    2.000000   
50%      0.000000   17.000000    1.000000    0.000000    3.000000    2.000000   
75%      1.000000   18.000000    1.000000    1.000000    4.000000    3.000000   
max      1.000000   22.000000    1.000000    1.000000    4.000000    4.000000   

       traveltime   studytime    failures   schoolsup  ...  guardian_mother  \
count  395.000000  395.000000  395.000000  395.000000  ...       395.000000   
mean     1.448101    2.035443    0.334177    0.129114  ...         0.691139   
std      0.697505    0.839240    0.743651    0.335751  ...         0.462610   
min      1.000000    1.000000    0.000000    0.000000  ...         0.000000   
25%      1.000000    1.000000    0.000000    0.000000  ...         0.000000   
50%      1.000000    2.000000    0.000000    0.000000  ...         1.000000   
75%      2.000000    2.000000    0.000000    0.000000  ...         1.000000   
max      4.000000    4.000000    3.000000    1.000000  ...         1.000000   

       guardian_other  reason_course  reason_home  reason_reputation  \
count      395.000000     395.000000   395.000000         395.000000   
mean         0.081013       0.367089     0.275949           0.265823   
std          0.273201       0.482622     0.447558           0.442331   
min          0.000000       0.000000     0.000000           0.000000   
25%          0.000000       0.000000     0.000000           0.000000   
50%          0.000000       0.000000     0.000000           0.000000   
75%          0.000000       1.000000     1.000000           1.000000   
max          1.000000       1.000000     1.000000           1.000000   

       Fjob_other  Fjob_services  Mjob_at_home  Mjob_other  Mjob_services  
count  395.000000     395.000000    395.000000  395.000000     395.000000  
mean     0.549367       0.281013      0.149367    0.356962       0.260759  
std      0.498188       0.450064      0.356902    0.479711       0.439606  
min      0.000000       0.000000      0.000000    0.000000       0.000000  
25%      0.000000       0.000000      0.000000    0.000000       0.000000  
50%      1.000000       0.000000      0.000000    0.000000       0.000000  
75%      1.000000       1.000000      0.000000    1.000000       1.000000  
max      1.000000       1.000000      1.000000    1.000000       1.000000  

[8 rows x 35 columns]

### Random Forest after Feature Reduction

In [46]:
rf = RandomForestClassifier() #Build
rf.fit(X_train,  y_train) #Train
y_predict = rf.predict(X_test)  #Make Prediction

<ipython-input-46-2b2288391eb4>:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(X_train,  y_train) #Train


In [47]:
print("Accuracy_score is:",round(accuracy_score(y_test,y_predict),2))

Accuracy_score is: 0.69


INSIGHT
* It is found that after carrying out the feature importance and feature reduction on random forest,then,rechecking on the algorithm, there is an increase in the value of accuracy score from 0.65 to 0.69. So this has increased the accuracy of the algorithm.

## Linear Regression after Standard Scaling

 ### Standard Scaling

In [48]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

In [49]:
lr = LogisticRegression() # Build 
lr.fit(X_train,  y_train) # Train
y_predict = lr.predict(X_test) # Make Prediction

C:\Users\Acer\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [50]:
print("Accuracy_score is:",round(accuracy_score(y_test,y_predict),2))

Accuracy_score is: 0.74


INSIGHT
* Here, it is found that the standard scaling carried out also has increased the accuracy score of linear regression from 0.70 to 0.74.Therefore the model has become more accurate. 